common/iterator/sorted/IntersectForwardableIterator.java

```diff
@@ -64,17 +76,43 @@ public boolean hasNext() {
         }
     }
 
-    private boolean fetch() {
+    private boolean computeWithinIntersection() {
+        assert state == State.EMPTY;
+        while (!intersectionIterators.isEmpty()) {
+            Forwardable<T, ORDER> iterator = intersectionIterators.get(0);
+            assert iterator.hasNext();
+            T next = iterator.peek();
+            if (isIntersection(candidate, next)) {
+                iterator.next();
+                if (!intersectionValues.contains(next)) {
+                    candidate = next;
+                    state = State.FETCHED;
+                    return true;
+                }
+                if (!iterator.hasNext()) intersectionIterators.remove(0);
+            } else {
+                intersectionIterators.remove(0);
+            }
+        }
+        intersectionValues.clear();
+        return false;
+    }
+
+    private boolean computeIntersection() {
+        assert state == State.INIT || state == State.EMPTY;
         Forwardable<T, ORDER> iterator = iterators.get(0);
         if (!iterator.hasNext()) state = State.COMPLETED;
         else {
-            candidateSource = 0;
```

Comment:
we replace an `int` index with the object it is indexing directly to simplify the understanding